glib/src/boxed_any_object.rs

```diff
@@ -0,0 +1,133 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate as glib;
+use crate::subclass::prelude::*;
+use crate::{object_subclass, wrapper, Object};
+use std::any::Any;
+use std::cell::{Ref, RefCell, RefMut};
+
+#[derive(thiserror::Error, Debug)]
+pub enum BorrowError {
+    #[error("Can't convert item inside BoxedAnyObject to requested type")]
+    InvalidType,
+    #[error("Item already immutably borrowed")]
+    AlreadyBorrowed(#[from] std::cell::BorrowError),
+    #[error("Item already mutably borrowed")]
+    AlreadyMutBorrowed(#[from] std::cell::BorrowMutError),
+}
+
+mod imp {
+    use super::*;
+
+    #[derive(Debug)]
+    pub struct BoxedAnyObject {
+        pub item: RefCell<Box<dyn Any>>,
+    }
+    #[object_subclass]
+    impl ObjectSubclass for BoxedAnyObject {
+        const NAME: &'static str = "BoxedAnyObject";
+        type Type = super::BoxedAnyObject;
+    }
+    impl Default for BoxedAnyObject {
+        fn default() -> Self {
+            Self {
+                item: RefCell::new(Box::new(None::<usize>)),
+            }
+        }
+    }
+    impl ObjectImpl for BoxedAnyObject {}
+}
+
+wrapper! {
+    /// # Examples
+    /// ```rust
+    /// use glib::BoxedAnyObject;
+    /// struct Author {
+    ///     name: String,
+    ///     subscribers: usize
+    /// }
+    /// // BoxedAnyObject can contain any custom type
+    /// let boxed = BoxedAnyObject::new(Author {
+    ///     name: String::from("GLibAuthor"),
+    ///     subscribers: 1000
+    /// });
+    ///
+    /// // The boxed data can be stored as a GObject
+    /// let list = gio::ListStore::new(BoxedAnyObject::static_type());
+    /// list.append(boxed.clone().upcast());
+    ///
+    /// // And can be retrieved with `borrow`
+    /// let author: Author = boxed.borrow();
+    /// ```
+    pub struct BoxedAnyObject(ObjectSubclass<imp::BoxedAnyObject>);
+}
+
+impl Default for BoxedAnyObject {
+    fn default() -> Self {
+        Object::new(&[]).expect("Failed to create BoxedAnyObject")
+    }
+}
+impl BoxedAnyObject {
+    pub fn new<T: 'static>(item: T) -> Self {
+        let obj = Self::default();
+        obj.replace(item);
+        obj
+    }
+
+    /// Replaces the wrapped value with a new one, returning the old value, without deinitializing either one.
+    pub fn replace<T: 'static>(&self, t: T) -> Box<dyn Any> {
+        self.impl_().item.replace(Box::new(t) as Box<dyn Any>)
+    }
+
+    /// Immutably borrows the wrapped value, returning none if the value is currently mutably borrowed or
+    /// the inner value has never been seet
```

Comment:
```suggestion
    /// the inner value has never been set
```

Reply:
And also please add a dot at the end of the sentence.